refactor!: PaginationRange<T> is now an alias to StreamRange<T>

devjgm opened the pull request:
Fixes: https://github.com/googleapis/google-cloud-cpp/issues/5537

**BREAKING CHANGE**

This change removes a constructor that not intended to be part of the public API from several "Range" types in the Pub/Sub and Spanner APIs. The affected types are:

  * `google/cloud/pubsub/subscription_admin_connection.h`
    * `using ListSubscriptionsRange = google::cloud::internal::PaginationRange`
    * `using ListSnapshotsRange = google::cloud::internal::PaginationRange`
  * `google/cloud/pubsub/topic_admin_connection.h`
    * `using ListTopicsRange = google::cloud::internal::PaginationRange`
    * `using ListTopicSubscriptionsRange = google::cloud::internal::PaginationRange`
    * `using ListTopicSnapshotsRange = google::cloud::internal::PaginationRange`
  * `google/cloud/spanner/database_admin_connection.h`
    * `using ListDatabaseRange = google::cloud::internal::PaginationRange`
    * `using ListBackupOperationsRange = google::cloud::internal::PaginationRange`
    * `using ListDatabaseOperationsRange = google::cloud::internal::PaginationRange`
    * `using ListBackupsRange = google::cloud::internal::PaginationRange`
  * `google/cloud/spanner/instance_admin_connection.h`
    * `using ListInstancesRange = google::cloud::internal::PaginationRange`
    * `using ListInstanceConfigsRange = google::cloud::internal::PaginationRange`

All normal usage of these types will continue to work, i.e., they will all continue to work as ranges. However, the following constructor was removed:

https://github.com/googleapis/google-cloud-cpp/blob/d7d3576c6533dc15bd656b74bed8fc73e7400980/google/cloud/internal/pagination_range.h#L142-L144



<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5538)
<!-- Reviewable:end -->

coryan reviewed (commented):
> Reviewed 18 of 19 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5538)* status: all files reviewed, 2 unresolved discussions (waiting on @devjgm)

___
*[google/cloud/internal/pagination_range.h, line 48 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5538#-MMuR6agDEgLO38rSTPp:-MMuR6agDEgLO38rSTPq:b94frci) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/56f0ea08f4d11e030b328f9a3ec8aff856df4e64/google/cloud/internal/pagination_range.h#L48)):*
> ```objc
>  */
> template <typename T>
> using PaginationRange = StreamRange<T>;
> ```

FYI, I think I can answer the question as to whether this is a breaking change. I think it is not: applications may have used the `internal::PaginationRange<T, Request, Response>` alias directly, but that requires accessing `internal`.

I considered using:

```cc
template <typename T, typename Request, typename Response>
using PaginationRange = StreamRange<T>;
```

but that is unnecessary.
___
*[google/cloud/internal/pagination_range.h, line 142 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5538#-MMuO7i109X2Uvs79HRv:-MMuO7i109X2Uvs79HRw:bue38a8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/56f0ea08f4d11e030b328f9a3ec8aff856df4e64/google/cloud/internal/pagination_range.h#L142)):*
> ```objc
> Range MakePaginationRange(Request request, Loader loader, Extractor extractor) {
>   using ValueType = typename Range::value_type::value_type;
>   using LoaderResult = typename std::result_of<Loader(Request)>::type;
> ```

consider using `google::cloud::internal::invoke_result_t<Loader, Request>`


<!-- Sent from Reviewable.io -->

devjgm commented:
> ```objc
> ```objc
>  */
> template <typename T>
> using PaginationRange = StreamRange<T>;
> ```
> 
> 
> 
> FYI, I think I can answer the question as to whether this is a breaking change. I think it is not: applications may have used the internal::PaginationRange<T, Request, Response> alias directly, but that requires accessing internal.
> ```

My main concern was about `PaginationRange`'s constructor:

coryan commented:
> > ```objc

Why does GitHub think I want to write Objective-C ? Nevermind..

> My main concern was about `PaginationRange`'s constructor:

Yeah, I saw #5539 after I wrote that.

> 
> Thoughts?

I think it is breaking, but we should make the change.  So you need to update `CHANGELOG.md`, this becomes a `refactor!:` change, and so on. But we still proceed.

devjgm commented:
> _[google/cloud/internal/pagination_range.h, line 142 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5538#-MMuO7i109X2Uvs79HRv:-MMuO7i109X2Uvs79HRw:bue38a8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/56f0ea08f4d11e030b328f9a3ec8aff856df4e64/google/cloud/internal/pagination_range.h#L142)):_
> 
> > ```objc
> > Range MakePaginationRange(Request request, Loader loader, Extractor extractor) {
> >   using ValueType = typename Range::value_type::value_type;
> >   using LoaderResult = typename std::result_of<Loader(Request)>::type;
> > ```
> 
> consider using `google::cloud::internal::invoke_result_t<Loader, Request>`

Good suggestion. I never before realized the problems with `std::result_of`. Changed.

devjgm commented:
> I think it is breaking, but we should make the change. So you need to update `CHANGELOG.md`, this becomes a `refactor!:` change, and so on. But we still proceed.

I think I agree. I added a note to `CHANGELOG.md`, and updated this PR description.

PTAL.

coryan reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5538#-:-MMuv8QBAp7HgVMvZEXX:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
